### PR TITLE
Network config

### PR DESCRIPTION
**Pull request by ryutamago**
# Context

We want to be able to pass a network option in the CLI and specify the options of networks in the config file.
# Description
The relevant cli commands takes the `--network, -n` option now.
* if the option not provided it will default to `default_network` in the config file.
* if the `default_network` isn't provided or the `--network` isn't aviable in the config file, it will throw
* we can set the `default_network` in the config file or `--network` in the CLI to `dev` to use the sandbox environment 
Please add the following network config
```
  "networks": {
    "default_network": null,
    "networks": {}
  }
  ```
# Manually testing the PR

https://github.com/trilitech/jstz/assets/128799322/08de604b-e2c3-4780-b2bb-0e533ef64220

### CODE REVIEW

File: crates/jstz_cli/src/account.rs [hunk 1]
```diff
@@ -216,14 +216,18 @@ fn list_accounts(long: bool) -> Result<()> {
...
 }
 
-async fn get_code(account: Option<AddressOrAlias>) -> Result<()> {
+async fn get_code(
+    account: Option<AddressOrAlias>,
+    network: Option<NetworkName>,
+) -> Result<()> {
     let cfg = Config::load()?;
 
+    println!("Getting code.. {:?}.", network);
```
Comment (johnyob) [inline crates/jstz_cli/src/account.rs:225]: Some of our commands are designed to be pipe-able (in the case of `jstz account code ... > index.js`)

```suggestion
    debug!("Getting code.. {:?}.", network);
```

File: crates/jstz_cli/src/config.rs [hunk 2]
```diff
@@ -194,10 +201,63 @@ pub struct SandboxConfig {
...
+
+impl<'de> Deserialize<'de> for NetworkName {
+    fn deserialize<D>(deserializer: D) -> core::result::Result<NetworkName, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s: String = Deserialize::deserialize(deserializer)?;
+        NetworkName::from_str(&s).map_err(serde::de::Error::custom)
+    }
+}
```
Comment (johnyob) [inline crates/jstz_cli/src/config.rs:248]: I'm fairly certain you can configure serde derive to serialize/deserialize based on `to/from_str`.

File: crates/jstz_cli/src/config.rs [hunk 3]
```diff
@@ -282,14 +342,71 @@ impl Config {
...
+            _ => None,
         })
     }
+
+    fn network(&self, name: &Option<NetworkName>) -> Result<Network> {
+        let network = match name {
+            Some(name) => self.lookup_network(name),
+            None => {
+                let name = self.networks.default_network.as_ref().ok_or_else(||user_error!(
+                    "No default network found in the config file. Please specify the <NETWORK> option or set the default network in the config file."
```
Comment (johnyob) [inline crates/jstz_cli/src/config.rs:381]:
```suggestion
                    "No default network found in the config file. Please specify the `--network` flag or set the default network in the config file."
```

File: crates/jstz_cli/src/logs/mod.rs [hunk 4]
```diff
@@ -14,9 +14,13 @@ pub enum Command {
         // The address or the alias of the deployed smart function.
         #[arg(value_name = "ALIAS|ADDRESS")]
         smart_function: AddressOrAlias,
-        // Optional log level to filter log stream.
+        // Optional log level to filter log stream
```
Comment (johnyob) [inline crates/jstz_cli/src/logs/mod.rs:17]: Revert, prefer fullstops are the end of docs for each flag

File: crates/jstz_cli/src/main.rs [hunk 5]
```diff
@@ -43,9 +43,13 @@ enum Command {
         /// Initial balance of the function.
         #[arg(short, long, default_value_t = 0)]
         balance: u64,
-        /// Name (or alias) of the function.
-        #[arg(short, long, default_value = None)]
+        /// Name
```
Comment (johnyob) [inline crates/jstz_cli/src/main.rs:46]: Prefer old doc, removing short option is fine though

File: crates/jstz_cli/src/sandbox/daemon.rs [hunk 6]
```diff
@@ -119,31 +120,31 @@ fn init_client(cfg: &Config) -> Result<()> {
     wait_for_node_to_initialize(cfg)?;
 
     // 2. Wait for the node to be bootstrapped
-    info!("Waiting for node to bootstrap...");
-    cfg.octez_client()?.wait_for_node_to_bootstrap()?;
-    info!(" done");
+    print!("Waiting for node to bootstrap...");
+    cfg.octez_client_sandbox()?.wait_for_node_to_bootstrap()?;
+    println!(" done");
```
Comment (johnyob) [inline crates/jstz_cli/src/sandbox/daemon.rs:126]: Please don't revert the logging changes (ditto else where). After rebasing on `main`, these should be `debug!`s

File: crates/jstz_cli/src/config.rs [hunk 7]
```diff
@@ -198,10 +206,63 @@ pub struct SandboxConfig {
...
+
+// impl<'de> Deserialize<'de> for NetworkName {
+//     fn deserialize<D>(deserializer: D) -> core::result::Result<NetworkName, D::Error>
+//     where
+//         D: Deserializer<'de>,
+//     {
+//         let s: String = Deserialize::deserialize(deserializer)?;
+//         NetworkName::from_str(&s).map_err(serde::de::Error::custom)
+//     }
+// }
```
Comment (johnyob) [inline crates/jstz_cli/src/config.rs:253]: Avoid commit commented code

**Review by johnyob — APPROVED** [review]
lgtm, tested locally. 

Please clean the commit history before merging

File: crates/jstz_cli/src/config.rs (hunk 7 above)
Comment (ryutamago) [reply crates/jstz_cli/src/config.rs:253]: nice catch, thanks :)

**Pull request merged**